text/2020-01-30-key-spaces.md

```diff
@@ -8,119 +8,115 @@ When multiple applications store data in TiKV, we need a clear pattern for avoid
-Authentication is necessary to provide any meaningful security for key spaces, but I can start a separate RFC for this.
+Authentication is necessary to provide any meaningful security for key spaces.
+
+This RFC will not discuss resource isolation.
+Key spaces will be an important tool for resource isolation to leverage.
 
 #### Client view
 
-When a client first connects to the TiKV system, it will register itself to get a key space.
-This happens in the metadata component (PD).
-All requests to the storage component (TiKV) will include the key space.
-The TiKV storage component will transparently convert the key space into a key prefix that TiKV uses for its operations.
+When a client first connects to the TiKV system, it will register itself to get a key space prefix.
+The client can register for additional key spaces, but normally one application will have one key space.
+
+All key spaces operations take place in the metadata component (PD) which stores a registry of key spaces.
+
+The TiKV component does not know about key spaces. All requests to the storage component (TiKV) will include the key space prefix.
+Please see [this separate proposal that helps with sending key prefixes to TiKV](https://github.com/tikv/rfcs/pull/56).
 
-The client can register for additional key spaces.
 
 #### Metadata operations
 
 POST /keyspaces
 
 requires:
 
+* key space name
 * application name
 
 optional:
 
-* key space name
 * description
-* prefix
 
-This API will return the key space object, which includes an ID. Key spaces have an ID to allow their name to be changed.
-Clients are recommended to provide a key space name, but the API can also auto-generate one.
+system generated:
 
-The usage of an application name as an identity may end up being a de-normalization of data in a future iteration when identity is added for authentication and there are identity ids.
-I suggest that we can add an identity id as an alternative field to an application name. Similarly, the output key space object can include the identity id.
+* prefix
 
-Addtional GET APIs will also return keyspace objects
+This API will return the key space object. The key space name is the id of the keyspace and cannot be changed.
+The usage of an application name as an identity may end up being a de-normalization of data in a future iteration when true identity with authentication is added.
+The prefix that is generated is a binary sequence. This will start with a single byte sequence: this optimizes for the fact that few TiKV installations will have more than 255 applications.
+The sequence starts with 0x01 and increment by 1. The 0x00 prefix will remain reserved.
+A byte sequence containing 0xFF for all bytes is not a valid prefix because this is the end of the range. When a sequence contains just 0xFF, a byte of 0x00 is appended and the byte sequence is now one byte longer. This pattern allows us to start with just a 1 byte sequence but allows expansion to any length.
```

Comment:
This is basically a varint encoding, it doesn' solve the problem that a key with shorter prefix may conflict with a key with longer prefix.

Reply:
Why not just use fixed 4 bytes, the simplicity benefit overweight the storage cost.

Reply:
yes this is similar to varint encoding. The difference is that varint encoding reserves an entire bit to indicate more length. This reserves only a single byte value (0xFF) to indicate more length.

Either technique does in fact prevent conflicts. I guess I need to more explicitly state that every 2 byte sequence generated by this pattern will start with 0xFF.

Reply:
I switched to the varint encoding